static/js/emoji_util.js

```diff
@@ -0,0 +1,114 @@
+import _ from "lodash";
+
+/*
+    This module was split out in early 2022 for mostly tactical
+    reasons. All the functions here are pure functions, which makes
+    them particularly amenable for code reuse with out mobile team.
+    I didn't move them directly to our shared library, because
+    it's not totally clear mobile will reuse them as currently
+    written.
+
+    This module is also a good candidate for converting to TypeScript.
+*/
+
+export function build_emoticon_translations({emoticon_conversions}) {
+    /*
+
+    Please keep this as a pure function so that we can
+    eventually share this code with the mobile codebase.
+
+    Build a data structure that looks like something
+    like this:
+
+    [
+        { regex: /(\:\))/g, replacement_text: ':smile:' },
+        { regex: /(\(\:)/g, replacement_text: ':smile:' },
+        { regex: /(\:\/)/g, replacement_text: ':confused:' },
+        { regex: /(<3)/g, replacement_text: ':heart:' },
+        { regex: /(\:\()/g, replacement_text: ':frown:' },
+        { regex: /(\:\|)/g, replacement_text: ':expressionless:' },
+        ....
+    ]
+
+        We build up this list of ~12 emoticon translations even
+        if user_settings.translate_emoticons is false, since
+        that setting can be flipped via live update events.
+        On the other hand, we assume that emoticon_conversions
+        won't change until the next reload, which is fine for
+        now (and we want to avoid creating new regexes on
+        every new message).
+    */
+
+    const translations = [];
+    for (const [emoticon, replacement_text] of Object.entries(emoticon_conversions)) {
+        const regex = new RegExp("(" + _.escapeRegExp(emoticon) + ")", "g");
+
+        translations.push({
+            regex,
+            replacement_text,
+        });
+    }
+
+    return translations;
+}
+
+export function build_emojis_by_name({
+    realm_emojis,
+    emoji_catalog,
+    get_emoji_name,
+    default_emoji_aliases,
+}) {
+    // Please keep this as a pure function so that we can
+    // eventually share this code with the mobile codebase.
+    const map = new Map();
+    for (const [realm_emoji_name, realm_emoji] of realm_emojis) {
+        const emoji_dict = {
+            name: realm_emoji_name,
+            display_name: realm_emoji_name,
+            aliases: [realm_emoji_name],
+            is_realm_emoji: true,
+            url: realm_emoji.emoji_url,
+            has_reacted: false,
+        };
+        map.set(realm_emoji_name, emoji_dict);
+    }
+
+    for (const codepoints of Object.values(emoji_catalog)) {
+        for (const codepoint of codepoints) {
+            const emoji_name = get_emoji_name(codepoint);
+            if (emoji_name !== undefined && !map.has(emoji_name)) {
```

Comment:
I think if we do that we'd need to reorder, since we're clearly trying to avoid overwriting realm emoji with unicode emoji. 

But I think removing this `has` and then swapping the order would make sense to me.

Reply:
Yup, that's what I had in mind.

Reply:
Makes sense, this will be in the next push.